### ISSUE TITLE
Multiple candidate class dictionaries

### ISSUE THREAD

**Issue by oxinabox** [google-research/dex-lang#340]
Continuing with from #339 

If i have a type-class
```
data Hashable a:Type = MkHashable (a->a->Bool) (Int->a->Int) -- isequal, hash
```

and I define a instance of it for `Int32`
```
@instance Int32Hashable : Hashable Int32 = (MkHashable
    (\a b. a==b)
    (\seed x. I64ToI $ hash (IToI64 seed) x)
)
```

and also a generic fallback defintion that work on anything that has `Eq`:
```
@instance FallbackHashable : Eq t ?=> Hashable t = (MkHashable
    (\a b. a==b)
    (\seed x. 1)  -- performance degrades to O(n) search 🤷
)
```

and then i try and use this:
```
tryfindlast : (Hashable a) ?=> (n=>a)->a->(Maybe n) = \col val.
    fold Nothing \i state. 
        select (col.i `isequal` val) (Just i) state

:p tryfindlast [1,2,3,2,3,4] 3
```
I get an error:
```
Type error:Multiple candidate class dictionaries for: (Hashable Int32)
```

I expected that it would chose the most specific instance.
Which I think loosely can be defined as the one with the fewest things on the left of the `?=>`.

it seems like being able to this would be useful so on can define general ways of doing things as well as optimized special cases.

**Comment by apaszke** [comment]
Yes, this is a pretty standard problem with type-classes. Haskell, for example, has the same restriction as we do, but adds an extension that lets you declare some type classes as preferable or less or via the `OVERLAPPING` pragmas.

It's a difficult problem in general, but I agree that we should at least to have a good way of solving this "provide a default" case.

**Comment by apaszke** [comment]
I think that some reason to be conservative here is that the behavior of your code might change if only you change imports/exports. Allowing duplicate classes might mean that `Hash Int` is one thing in one module (e.g. an optimized implementation), but something completely different in another one, where you forgot to include the module with the optimized class.

**Comment by oxinabox** [comment]
> I think that some reason to be conservative here is that the behavior of your code might change if only you change imports/exports. Allowing duplicate classes might mean that Hash Int is one thing in one module (e.g. an optimized implementation), but something completely different in another one, where you forgot to include the module with the optimized class.

Julia has a similar problem, which you can find information about by looking up Type Piracy (which is not forbidden by language but is strongly recommended against. But could easily enough be forbidden by compiler).

The equivelent solution for type classes would be:
A module can only declare a `instance` if it either:
 - Declared the typeclass this is an instance of
 - Or: declared one or more of the types that the instance is applying to.

So the module that declared `Hashable` is the only one that is allowed to give very general definations that can apply to any `Type` or to `Int` or any of hte other prelude types.
And for prelude type classes like `Ord`, all instances for prelude types (or that are general enough to accept prelude types) also must be declared in the prelude.

**Comment by apaszke** [comment]
Yes, that does sound reasonable. FWIW even Haskell with `OverlappingInstances` warns of instances that are defined in files that also don't define the data types (those are called orphan instances). I think that the approach you outline might be a good choice for Dex.

There's one other aspect to this, which is to say that we want the _default_ typeclass search mechanism to be conservative and make it hard for users to shoot themselves in the foot. But, we should still enable them to define instances which do not participate in the default search, with the caveat that those have to be explicitly specified (as an explicit specification of an implicit argument --- we're still missing that bit of syntax) at each use site.

**tl;dr** orphan instances are ok for as long as you're entirely explicit about their use, default mechanisms should be restricted

**Comment by danieldjohnson** [comment]
Purescript is a neat point in the design space for this, they allow "instance chains" that are searched over in a specific order: https://github.com/purescript/documentation/blob/master/language/Type-Classes.md#instance-chains

**Comment by apaszke** [comment]
That's really cool! But how does that play with extensibility?

**Comment by oxinabox** [comment]
> A module can only declare a `instance` if it either:
> 
>     * Declared the typeclass this is an instance of
> 
>     * Or: declared one or more of the types that the instance is applying to.

I just learned that rust has this rule for traits, it calls it the _orphan rule_.
https://doc.rust-lang.org/book/ch10-02-traits.html
Traits, Type-classes and Open Methods; are really 3 related methods to solve approximately the same problem.
